File: src/day4/main.cpp
/**
 * Yuck
 */

#include <algorithm>
#include <fstream>
#include <iostream>
#include <unordered_map>
#include <unordered_set>
#include <vector>

typedef std::vector<std::unordered_map<std::string, std::string>> choo;

typedef bool (*myfunc)(std::string x);

bool isfourDigits(std::string x) { return x.size() == 4; }
bool inrange(int x, int l, int m) { return x >= l && x <= m; };
bool isvalidheight(std::string x) {
  std::string ty = x.substr(x.length() - 2, x.length());
  std::string val = x.substr(0, x.length() - 2);
  if (ty != "cm" && ty != "in") {
    return false;
  };
  if (ty == "cm") {
    return inrange(std::stoi(val), 150, 193);
  } else {
    return inrange(std::stoi(val), 59, 76);
  };
};
bool validHairColor(std::string x) {
  std::string vals = x.substr(1, x.length());
  if (x[0] != '#' || vals.length() != 6) {
    return false;
  };
  if (std::all_of(vals.begin(), vals.end(), ::isxdigit)) {
    return true;
  };
  return false;
};

/**
 * A better way would be to pass requirements as functions (i.e. pass the
 * functions that return the valid conditions as arguments) Then just loop
 * through each argument and run each function and update the valid accordingly
 */
 template<typename... Args> bool f(Args &&...args);
struct funco {
  bool valid = true;
 bool checker(std::forward<f>(args)...) {
    valid = valid && args("test");
    return valid;
  };
};
bool is_number(const std::string &s) {
  return !s.empty() && std::find_if(s.begin(), s.end(), [](unsigned char c) {
                         return !std::isdigit(c);
                       }) == s.end();
}

/**
 * This is extremely gross
 */
int part2(choo passports, std::vector<std::string> list) {
  int valid = passports.size();
  for (auto i : passports) {
    for (auto k : list) {
      if (i.find(k) == i.end() && k != "cid") {
        valid--;
        break;
      }
      // DONE
      if (k == "cid")
        continue;
      // DONE
      else if (k == "byr") {
        // Not 4digits
        if (!isfourDigits(i[k])) {
          valid--;
          break;
        } else {
          // Not In range
          if (!inrange(std::stoi(i[k]), 1920, 2002)) {
            valid--;
            break;
          };
        }
      } else if (k == "eyr") {
        // Not 4 Dig
        if (!isfourDigits(i[k])) {
          valid--;
          break;
        } else {
          // Not in range
          if (!inrange(std::stoi(i[k]), 2020, 2030)) {
            valid--;
            break;
          };
        }
      } else if (k == "iyr") {
        // Not four dig
        if (!isfourDigits(i[k])) {
          valid--;
          break;
        } else {
          // Not in range
          if (!inrange(std::stoi(i[k]), 2010, 2020)) {
            valid--;
            break;
          };
        }
      } else if (k == "hgt") {
        // Not valid
        if (!isvalidheight(i[k])) {
          valid--;
          break;
        };
      } else if (k == "hcl") {
        // Not valid
        if (!validHairColor(i[k])) {
          valid--;
          break;
        }
      } else if (k == "ecl") {
        // Not in et
        std::unordered_set<std::string> types = {"amb", "blu", "brn", "gry",
                                                 "grn", "hzl", "oth"};
        if (types.find(i[k]) == types.end()) {
          valid--;
          break;
        }
      } else {
        // Not number or its length != 9
        if (!is_number(i[k]) || i[k].length() != 9) {
          valid--;
          break;
        }
      };
    };
  };
  return valid;
};

void loadEntry(std::string en,
               std::unordered_map<std::string, std::string> &map) {
  int colon = en.find(":");
  map[en.substr(0, colon)] = en.substr(colon + 1, en.length() - 1);
};

void loadIt(choo &input) {
  std::ifstream file("./input.txt");
  if (file.is_open()) {
    std::string entry;
    int pp = 0;
    std::unordered_map<std::string, std::string> test;
    input.emplace_back(test);
    while (std::getline(file, entry)) {
      if (entry.empty()) {
        pp++;
        std::unordered_map<std::string, std::string> test2;
        input.emplace_back(test2);
      } else {
        while (entry.length() != 0) {
          int spaceDelim = entry.find(" ");
          if (spaceDelim != std::string::npos) {
            std::string s = entry.substr(0, spaceDelim);
            loadEntry(s, input[pp]);
            entry.erase(0, spaceDelim + 1);
          } else {
            loadEntry(entry, input[pp]);
            entry.erase(0, entry.length());
          }
        };
      }
    };
    std::cout << input.size() << std::endl;
  };
}

int part1(choo passports, std::vector<std::string> list) {
  int valid = passports.size();
  for (auto i : passports) {
    for (auto k : list) {
      if (k == "cid")
        continue;
      if (i.find(k) == i.end()) {
        valid--;
        break;
      }
    };
  };
  return valid;
}

int main() {
  choo passports{};
  std::vector<std::string> list = {"byr", "iyr", "eyr", "hgt",
                                   "hcl", "ecl", "pid", "cid"};
  loadIt(passports);

  std::cout << "Passports : " << passports.size() << std::endl;

  std::cout << "Valid :" << part1(passports, list) << std::endl;
  std::cout << "Crazy :" << part2(passports, list) << std::endl;
};

File: go/day4/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

func reader(path string) ([]string, error) {
	file, er := os.Open(path)

	if er != nil {
		return nil, er
	}
	defer file.Close()

	var inputs []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		if scanner.Text() != "" {
			inputs = append(inputs, scanner.Text())
		}
	}
	return inputs, scanner.Err()
}

func isInRange(x int, l int, h int) bool { return x >= l && x <= h }
func isFourDigits(x string) bool         { return len(x) == 4 }

var ppscanner = map[string]func(string) bool{
	"byr": func(x string) bool {
		value, err := strconv.Atoi(x)
		if err != nil {
			return false
		}
		return (isFourDigits(x) && isInRange(value, 1920, 2002))
	},
	"iyr": func(x string) bool {
		value, err := strconv.Atoi(x)
		if err != nil {
			return false
		}
		return (isFourDigits(x) && isInRange(value, 2010, 2020))
	},
	"eyt": func(x string) bool {
		value, err := strconv.Atoi(x)
		if err != nil {
			return false
		}
		return (isFourDigits(x) && isInRange(value, 2020, 2030))
	},
	"hgt": func(x string) bool {
		//Get in or CM
		cm := strings.Index(x, "cm")
		in := strings.Index(x, "in")
		if cm == -1 && in == -1 {
			fmt.Println("Not found")
			return false
		}
		if in != -1 {
			inch_value, err := strconv.Atoi(x[:in])
			if err != nil {
				return false
			}
			return inch_value >= 59 && inch_value <= 76
		} else {
			cm_value, err := strconv.Atoi(x[:in])
			if err != nil {
				return false
			}
			return cm_value >= 150 && cm_value <= 76
		}
	},
	"hcl": func(x string) bool {
		if len(x) != 7 {
			return false
		}
		//check hex
		if x[0] != '#' {
			return false
		}
		reg, err := regexp.Compile(`[0-9a-f]*`)
		if err != nil {
			return false
		}
		if !reg.MatchString(x[1:]) {
			return false
		}
		return true
	}, "ecl": func(x string) bool {
		eyemap := map[string]bool{"amb": true, "blu": true, "brn": true, "gry": true, "grn": true, "hzl": true, "oth": true}
		if ok := eyemap[x]; ok {
			return true
		}
		return false
	}, "pid": func(x string) bool {
		regex, _ := regexp.Compile(`^\d{9}$`)
		return regex.MatchStrring(x)

	}, "cid": func(x string) bool {
		return true
	},
}

func main() {

	inputs, _ := reader("test.txt")

}
